src/bluetooth_obex.rs

```diff
@@ -0,0 +1,165 @@
+extern crate dbus;
+use self::dbus::arg::{Dict, Variant};
+use self::dbus::Path as ObjectPath;
+use self::dbus::{BusType, Connection, Message, MessageItem, Props};
+use std::collections::HashMap;
+use std::error::Error;
+use std::path::Path;
+use std::thread::sleep;
+use std::time::Duration;
+
+use bluetooth_device::BluetoothDevice;
+
+const OBEX_BUS: &str = "org.bluez.obex";
+const OBEX_PATH: &str = "/org/bluez/obex";
+const OBJECT_PUSH_INTERFACE: &str = "org.bluez.obex.ObjectPush1";
+const CLIENT_INTERFACE: &str = "org.bluez.obex.Client1";
+const TRANSFER_INTERFACE: &str = "org.bluez.obex.Transfer1";
+
+pub enum SessionTarget {
+    Ftp,
+    Map,
+    Opp,
+    Pbap,
+    Syncr,
+}
+
+impl SessionTarget {
+    fn as_str(&self) -> &str {
+        match self {
+            SessionTarget::Ftp => "ftp",
+            SessionTarget::Map => "map",
+            SessionTarget::Opp => "opp",
+            SessionTarget::Pbap => "pbap",
+            SessionTarget::Syncr => "sync",
+        }
+    }
+}
+
+pub enum TransferState {
+    Queued,
+    Active,
+    Complete,
+    Suspended,
+    Error,
+}
+
+impl TransferState {
+    fn as_str(&self) -> &str {
+        match self {
+            TransferState::Queued => "queued",
+            TransferState::Active => "active",
+            TransferState::Complete => "complete",
+            TransferState::Suspended => "suspended",
+            TransferState::Error => "error",
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct BluetoothOBEXSession {
+    connection: Connection,
+    object_path: String,
+}
+
+impl BluetoothOBEXSession {
+    // https://git.kernel.org/pub/scm/bluetooth/bluez.git/tree/doc/obex-api.txt#n12
+    pub fn new(
+        connection: Connection,
+        device: &BluetoothDevice,
+    ) -> Result<BluetoothOBEXSession, Box<Error>> {
+        let device_address: String = device.get_address()?;
+        let mut map = HashMap::new();
+        map.insert("Target", Variant(SessionTarget::Opp.as_str()));
+        let args: Dict<&str, Variant<&str>, _> = Dict::new(map);
+        let m = Message::new_method_call(OBEX_BUS, OBEX_PATH, CLIENT_INTERFACE, "CreateSession")?
+            .append2(device_address, args);
+
+        let r = connection.send_with_reply_and_block(m, 1000)?;
+        let session_path: ObjectPath = r.read1()?;
+        let session_str: String = session_path.parse()?;
+        let obex_session = BluetoothOBEXSession {
+            connection,
+            object_path: session_str,
+        };
+        Ok(obex_session)
+    }
+
+    // https://git.kernel.org/pub/scm/bluetooth/bluez.git/tree/doc/obex-api.txt#n35
+    pub fn remove_session(&self) -> Result<(), Box<Error>> {
+        Message::new_method_call(OBEX_BUS, OBEX_PATH, CLIENT_INTERFACE, "RemoveSession")?
```

Comment:
I can see that you create a call, but you didn't send it via connection.
Or maybe I missing something?

Reply:
Session is referenced by the object path, which is enough for D-Bus to identify which resource (session) to deactivate, so this should be fine.